### PR TITLE
Refactored scheduler callbacks (epoch-based/step-based warmup)

### PR DESCRIPTION
**Pull request by BloodAxe**
## Changes:

`warmup_mode: linear_step` will emit deprecation warning. But will continue to work without any changes.

Two new modes with explicit meaning:

* `warmup_mode: linear_batch_step` 
* `warmup_mode: linear_epoch_step`

## Per-batch warmup:

```
warmup_mode: linear_batch_step
num_warmup_steps: 100
```

## Old Per-epoch warmup:

```
warmup_mode: linear_batch_step
num_warmup_epochs: 4
```

### CODE REVIEW

File: src/super_gradients/training/utils/callbacks/callbacks.py [hunk 1]
```diff
@@ -398,6 +398,60 @@ def is_lr_scheduling_enabled(self, context):
...
+
+    def perform_scheduling(self, context):
+        global_training_step = context.batch_idx + context.epoch * self.train_loader_len
+        self.lr = self.learning_rates[global_training_step]
+        self.update_lr(context.optimizer, context.epoch, None)
+        logger.debug(f"Setting learning rate to for {self.lr:e} at warmup step {global_training_step}")
+
+    def is_lr_scheduling_enabled(self, context):
+        global_training_step = context.batch_idx + context.epoch * self.train_loader_len
+        return global_training_step <= self.lr_warmup_steps
```
Comment (shaydeci) [inline src/super_gradients/training/utils/callbacks/callbacks.py:452]: I also this would be inaccurate (but this also comes from the phase mistake I addressed before...).
I think that if, for example, we have just 1 warmup epoch, then in the second epoch we still don't reach the initial lr, up until the first batch.

This isn't super critical, but just mildly inaccurate. This would also result in confusing lr logging in the TBs.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Nice work (: 
There are some comments inline., mostly discussing some edge cases and mildly (possible) inaccuracies.

Its important IMO we add unit tests for scheduling. Please see the different use cases I had in tests/unit_tests/lr_warmup_test.py.

File: src/super_gradients/training/utils/callbacks/callbacks.py [hunk 2]
```diff
@@ -398,6 +398,60 @@ def is_lr_scheduling_enabled(self, context):
...
+            f"Instantiating LinearWarmupLRCallback. "
+            f"Warmup scaling LR from {warmup_initial_lr:e} to {initial_lr:e} across {lr_warmup_steps} steps / {lr_warmup_steps} / {train_loader_len} epoch(s))."
+        )
+
+        self.learning_rates = learning_rates
+        self.train_loader_len = train_loader_len
+        self.lr_warmup_steps = lr_warmup_steps
+
+    def perform_scheduling(self, context):
+        global_training_step = context.batch_idx + context.epoch * self.train_loader_len
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/utils/callbacks/callbacks.py:351]: So the step-based warmup logic has been simplified to:
a) Avoid interference with any epoch-based schedulers by hard-capping max number of warmup steps to not exceed one epoch. With this change warmup cannot exceed one epoch and it eliminate any clash with cosine decay LR for instance.

b) When gradient accumulation is enabled, the LR change will happen only when gradient step is about to happen (e.g before optimizer.step). The number of warmup steps remains the same.
By overriding `on_train_batch_gradient_step_start` method we don't have to manually track total number of steps & accumulation counter.

File: src/super_gradients/training/utils/callbacks/callbacks.py [hunk 3]
```diff
@@ -398,6 +398,60 @@ def is_lr_scheduling_enabled(self, context):
...
+
+        if lr_warmup_epochs is not None and lr_warmup_steps is not None:
+            logger.warning(
+                f"Both lr_warmup_epochs and lr_warmup_steps arguments are passed to LinearWarmupLRCallback. "
+                f"A lr_warmup_steps({lr_warmup_steps}) will overtake. "
+                f"To remove this warning please remove either `lr_warmup_steps` or `lr_warmup_epochs` from your config file."
+            )
+
+        if lr_warmup_steps is None:
+            lr_warmup_steps = int(train_loader_len * lr_warmup_epochs)
```
[the remark this answers is not in the document]
Comment (BloodAxe) [reply src/super_gradients/training/utils/callbacks/callbacks.py]: With a given design of warmup schedulers we don't have info what scheduling logic is following warmup so it is just impossible to account. 
So my suggestion is to cap the max number of warmup steps to not exceed the train loader length.

File: tests/unit_tests/lr_warmup_test.py [hunk 4]
```diff
@@ -110,6 +110,47 @@ def test_lr_warmup_with_lr_scheduling(self):
         # THE LRS AFTER THE UPDATE
         self.assertListEqual(lrs, expected_lrs)
 
+    def test_lr_warmup_per_step_with_lr_scheduling(self):
+        # Define model
+        net = LeNet()
+        trainer = Trainer("lr_warmup_test_per_step")
+
+        lrs = []
+        phase_callbacks = [TestLRCallback(lr_placeholder=lrs)]
```
Comment (shaydeci) [inline tests/unit_tests/lr_warmup_test.py:119]: I think it would be better to have another callback that would collect the lrs after batch_step (or alternatively at batch start) so we know the warmup actually happened.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Looks awesome!
I just have one final suggestion for the unit test, see my comment there.

**Review by shaydeci — APPROVED** [review]
LGTM
